core/schema/cache.go

```diff
@@ -19,8 +20,9 @@ func (s *cacheSchema) Name() string {
 
 func (s *cacheSchema) Install() {
 	dagql.Fields[*core.Query]{
-		dagql.Func("cacheVolume", s.cacheVolume).
+		dagql.NodeFunc("cacheVolume", s.cacheVolume).
 			Doc("Constructs a cache volume for a given cache key.").
+			Impure("cacheVolume should be namespaced").
```

Comment:
Saw that the test workflows in this PR seems noticeably slower and am wondering if the impurity here is the cause. Marking this as impure and then returning another instance that includes `cacheVolume` might be tainting too many dagql IDs and thus giving us less dagql cache benefit (I am just going off theory here though, worth verification).

If so, one solution would be to go back to having a separate `__internalCacheVolume` API that's not impure; this use case is just sort of different than some of the other recent ones where we managed to avoid that hack...

I'd really like it if we could find something better though since that hack really pollutes telemetry, the API, etc.

I think the fundamental problem is that the current impurity system is too coarse (something that's come up in many discussions, not a new idea). Fully reworking it is too much to block this PR on but trying to think of some relatively quick steps we could take to improve it and fix this problem.

I guess what we want is that impurity can be decided on per-call rather than fixed in stone in the schema spec. I actually think that's doable in terms of the low-level dagql Call format. So the basic idea would be that when we check for impurity [here](https://github.com/rajatjindal/dagger/blob/16be68ae06badbcc65e2a6c14aad3b4752035831/dagql/server.go#L891-L891), the impurity value in the spec is just a default but could be overridden.

*Maybe* this is just a matter of setting up that plumbing between a `dagql.Select` call and that line linked above?

cc @vito in case you have any ideas

@rajatjindal this is a bit of ranty comment, sorry. Let me know if it makes sense, can try to clarify more if needed.

Reply:
Took a swing at adding support for dynamic purity in DagQL - it's a PR that targets _this_ PR, so if we're good with the change we can either just merge it into this, or I can create a separate PR. It was just easiest to test in the context of the thing that needed it.

* https://github.com/rajatjindal/dagger/pull/1

tl;dr: now you can call `dagql.Taint(ctx)` dynamically anywhere in a resolver code path and it'll prevent the current API call from being cached, and also change the returned ID to be tainted - exactly the same effects of calling `Impure` at the schema level, just determined at runtime instead.

Reply:
Hi @vito

thanks for adding the support for dynamic purity. I rebased this branch on top of your changes (the new PR to the dagger repo) and accomodated the namespacing logic in this PR accordingly. It seems to have worked as expected.

thanks again.

Reply:
@rajatjindal Ah cool, glad it worked! Feel free to toss out some of the later commits, the socket + regen ones in particular - I think I need to do a bit more work to fix that case, but the rest of the PR seems OK.

Reply:
Hi @sipsma, one more thing I noticed, while testing these changes more, is that if we change the source of the module, that would mean the digest is changed and with the new `taint` functionality we will be creating a new cache volume. I don't think we want that behavior. 

I added a testcase for it, and it seems like that is indeed the case. if we use just the name for namespacing, that might not be sufficient too (high chance of namespace conflict). 

what do you think?

Reply:
@rajatjindal great point... it's definitely a delicate balance in terms of exactly how stable an ID we want to use for the module here.

One important point is that today, it's "extremely stable" in that it's not even tied to the module at all, so I think it would make sense for us to start out with something pretty stable still rather than suddenly switching to something that's changing all the time.

I think this is what you said you switched to in discord, but to confirm, I'm imagining:
1. For git modules, the module name + the URL (excluding scheme like `https`/`ssh`/etc.) + the module source path
2. For local modules, the module name + the module source path

This is definitely not flawless, but seems like an okay starting place. This is worth revisiting though. Do you mind opening an issue for discussing this point more so we don't lose track of it after merging this PR?

Reply:
I've reported https://github.com/dagger/dagger/issues/9017 as a follow up. I was not using module source path explicitly (as it is part of modPath), but I've added it as part of namespace key now. 👍